### PR TITLE
Add ERC: ERC-20 Transfer Reference Extension

### PR DESCRIPTION
**Pull request by radeksvarz**

### CODE REVIEW

File: ERCS/erc-xyz.md [hunk 1]
```diff
@@ -0,0 +1,166 @@
+---
+eip: XXXX
```
Comment (abcoathup) [inline ERCS/erc-xyz.md:2]:
```suggestion
eip: 7699
```
Assigning next sequential EIP/ERC/RIP number.
Numbering changed to sequential from 7500.

Please also update the filename.

File: ERCS/erc-xyz.md [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+title: ERC20 Payment reference extension
+description: A minimal standard interface for ERC20 tokens allowing users to include a unique identifier (payment reference) for each transaction to help distinguish and associate payments with orders and invoices.
+author: Radek Svarz (@radeksvarz)
+discussions-to: https://ethereum-magicians.org/t/erc-xxxx-erc20-payment-reference-extension/19826
```
Comment (abcoathup) [inline ERCS/erc-xyz.md:6]:
```suggestion
discussions-to: https://ethereum-magicians.org/t/erc-7699-erc20-payment-reference-extension/19826
```

**Review by Gemk83 — APPROVED** [review]

**Review by SamWilsn — COMMENTED** [review]
You'll need to fix all the errors generated by the linter: https://github.com/ethereum/ERCs/actions/runs/8858664958/job/24327583845

**Comment by radeksvarz** [comment]
> You'll need to fix all the errors generated by the linter: https://github.com/ethereum/ERCs/actions/runs/8858664958/job/24327583845

@SamWilsn Thx for letting me know. I will fix those. Would you be so kind and help initiate the discussion on related ethereum magicians among wallets (mainly Metamask, Rabby) and explorers (Etherscan, Blockscout)?

**Comment by SamWilsn** [comment]
@radeksvarz we'd be happy to have you on the next AllWalletDevs call! You can find the details and discord server here: https://allwallet.dev/

File: ERCS/erc-7699.md [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+## Backwards Compatibility
+
+This extension is fully backwards compatible with the existing ERC-20 token standard. The new functions can be used alongside the existing transfer and transferFrom functions. Existing upgradable ERC-20 tokens can be upgraded to include the new functionality without impact on the storage layout; new ERC-20 tokens can choose to implement the payment reference features based on their specific needs.
+
+ERC-20 requires its `Transfer(address indexed _from, address indexed _to, uint256 _value) ` event to be emitted during transfers, thus there will be duplicitous data logged (from, to, amount) in two events.
+
+## Reference Implementation
+
+```
+// SPDX-License-Identifier: MIT
```
Comment (SamWilsn) [inline ERCS/erc-7699.md:71]: I can't merge this until either: this license is changed to CC0-1.0, or the code is moved into the `../assets/erc-7699/` directory.

**Review by SamWilsn — CHANGES_REQUESTED** [review]

**Review by fhar28 — APPROVED** [review]

File: ERCS/erc-7699.md (hunk 3 above)
Comment (radeksvarz) [reply ERCS/erc-7699.md]: Thx, updated

**Comment by radeksvarz** [comment]
I added the section ## Privacy Considerations (reflecting the Web3Privacy Now initiative). Is that a problem?

**Comment by radeksvarz** [comment]
> I added the section ## Privacy Considerations (reflecting the Web3Privacy Now initiative). Is that a problem?

@SamWilsn I fixed the license, can you please react on the section issue?

**Comment by radeksvarz** [comment]
@abcoathup , @fhar28 pls help

I added the section ## Privacy Considerations (reflecting the Web3Privacy Now initiative). Is that a problem?

File: ERCS/erc-7699.md [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+event TransferReference(bytes32 indexed loggedReference);
+
+}
+```
+
+These `transfer` and `transferFrom` functions, in addition to the standard transfer behaviour, MUST emit a `transferReference` event with a `loggedReference` parameter (with only exception defined below).
+
+The corresponding ERC-20 `Transfer` event MUST be emitted following the `TransferReference` event, ideally immediately afterwards for the client to be able to seek the associated `Transfer` event log record. 
+
+Emitted `loggedReference` MIGHT be the exact copy of the `transferReference` (when less then 33 bytes) or the derived data from the rich `transferReference` structure and other processing. This is up to the implementer. One MUST NOT expect the `transferReference` and `loggedReference` data are equal.
```
Comment (SamWilsn) [inline ERCS/erc-7699.md:52]: Is "MIGHT" one of the keywords? I don't actually know.

Comment (SamWilsn) [inline ERCS/erc-7699.md:52]:
```suggestion
Emitted `loggedReference` MIGHT be the exact copy of the `transferReference` (when less then 33 bytes) or the derived data from the rich `transferReference` structure and other processing. This is up to the implementer. One MUST NOT expect the `transferReference` and `loggedReference` data to be equal.
```

File: ERCS/erc-7699.md [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+
+The `loggedReference` parameter MAY contain any data of bytes32 type.
+
+The `transferReference` parameter MAY be empty. In this case and only in this case the `TransferReference` event MUST NOT be emitted, effectively mimicking the regular ERC-20 transfer without any transfer reference. 
+
+The `transferReference` parameter is not limited in length by design, users are motivated to keep it short due to calldata and execution gas costs.
+
+The `TransferReference` event MUST NOT be anonymous. This is to ensure that the event signature is logged and can be used as a filter.
```
Comment (SamWilsn) [inline ERCS/erc-7699.md:60]: This _might_ be clear to someone reading this slower than I am, but I feel "anonymous" here is a bit vague. Might be something to fix while in Draft.

File: ERCS/erc-7699.md [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+    }
+}
+
+```
+
+## Security Considerations
+
+### Privacy Considerations
+
+Reference data privacy: Including payment references in token transfers may expose sensitive information about the transaction or the parties involved. Implementers and users SHOULD carefully consider the privacy implications and ensure that payment references do not reveal sensitive information. To mitigate this risk, implementers can consider using encryption or other privacy-enhancing techniques to protect payment reference data.
```
Comment (SamWilsn) [inline ERCS/erc-7699.md:198]:
```suggestion
Reference data privacy: Including payment references in token transfers may expose sensitive information about the transaction or the parties involved. Implementers and users should carefully consider the privacy implications and ensure that payment references do not reveal sensitive information. To mitigate this risk, implementers can consider using encryption or other privacy-enhancing techniques to protect payment reference data.
```

Avoid using the keywords outside the specification section.

**Review by SamWilsn — APPROVED** [review]
Feel free to fix these at some point while you're in draft.

File: ERCS/erc-7699.md (hunk 4 above)
Comment (radeksvarz) [reply ERCS/erc-7699.md:52]: @SamWilsn The keyword is "MAY" . Should I create a separate PR for these fixes , or edit this one?

File: ERCS/erc-7699.md (hunk 5 above)
Comment (radeksvarz) [reply ERCS/erc-7699.md:60]: ah, it refers to the type of event, resp. to the specifier as documented here: https://docs.soliditylang.org/en/latest/contracts.html#events

Would this be better?
"""
The `TransferReference` event MUST NOT be declared with the `anonymous` specifier.
"""

File: ERCS/erc-7699.md (hunk 4 above)
Comment (radeksvarz) [reply ERCS/erc-7699.md:52]: fixed with #716  Thx

Comment (radeksvarz) [reply ERCS/erc-7699.md:52]: fixed with #716  Thx

File: ERCS/erc-7699.md (hunk 5 above)
Comment (radeksvarz) [reply ERCS/erc-7699.md:60]: fixed with #716  Thx

File: ERCS/erc-7699.md (hunk 6 above)
Comment (radeksvarz) [reply ERCS/erc-7699.md:198]: fixed with #716  Thx

**Pull request merged**